topdown/jsonschema.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/internal/gojsonschema"
+)
+
+// astValueToJSONSchemaLoader converts a value to JSON Loader.
+// Value can be ast.String or ast.Object.
+func astValueToJSONSchemaLoader(value ast.Value) (gojsonschema.JSONLoader, error) {
+	var loader gojsonschema.JSONLoader
+	var err error
+
+	// ast.Value type selector.
+	switch x := value.(type) {
+	case ast.String:
+		// In case of string pass it as is as a raw JSON string.
+		// Make pre-check that it's a valid JSON at all because gojsonschema won't do that.
+		if !json.Valid([]byte(x)) {
+			return nil, errors.New("invalid JSON string")
+		}
+		loader = gojsonschema.NewStringLoader(string(x))
+	case ast.Object:
+		// In case of object serialize it to JSON representation.
+		var data interface{}
+		data, err = ast.JSON(value)
+		if err != nil {
+			return nil, err
+		}
+		loader = gojsonschema.NewGoLoader(data)
+	default:
+		// Any other cases will produce an error.
+		return nil, errors.New("wrong type, expected string or object")
+	}
+
+	return loader, nil
+}
+
+func newResultTerm(valid bool, data *ast.Term) *ast.Term {
+	return ast.ArrayTerm(ast.BooleanTerm(valid), data)
+}
+
+// builtinJSONSchemaValidate accepts 1 argument which can be string or object and checks if it is valid JSON schema.
+// Returns array [false, <string>] with error string at index 1, or [true, ""] with empty string at index 1 otherwise.
+func builtinJSONSchemaValidate(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Take first argument and make JSON Loader from it.
+	loader, err := astValueToJSONSchemaLoader(operands[0].Value)
+	if err != nil {
+		return iter(newResultTerm(false, ast.StringTerm("jsonschema: "+err.Error())))
+	}
+
+	// Check that schema is correct and parses without errors.
+	if _, err = gojsonschema.NewSchema(loader); err != nil {
+		return iter(newResultTerm(false, ast.StringTerm("jsonschema: "+err.Error())))
```

Comment:
Instead of returning a string here, what if we return an object with an `error` field describing the error similar to what `http.send` does?

Reply:
Now, we're returning an array with a boolean and an error string, similar (but kinda reversed) to what [graphql.parse_and_verify](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-graphql-graphqlparse_and_verify) does. E.g.
```rego
deny[err] {
   [false, err] = jsonschema.verify(input.schema)
}
```
We could return an object instead, if that is preferable.

Reply:
This seems fine. My reasoning behind using an object was it gives us flexibility to add more fields if needed.